### PR TITLE
D8NID-147 contacts site section

### PR DESCRIPTION
**Pull request by johangant**

### CODE REVIEW

**Comment by johangant** [comment]
To fix:

```
➜  drupal8 git:(D8NID-147-contact-listing) ✗ lando phpcs
Running coding standard checks in /app/drupal8/web/modules/custom /app/drupal8/web/modules/migrate /app/drupal8/web/themes/custom /app/drupal8/web/profiles/custom
Using config file: /app/drupal8/vendor/squizlabs/php_codesniffer/CodeSniffer.conf

Config value "installed_paths" updated successfully; old value was "/app/drupal8/vendor/drupal/coder/coder_sniffer"

FILE: /app/drupal8/web/modules/custom/nidirect_contacts/src/Plugin/Block/ContactAzBlock.php
-------------------------------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 4 LINES
-------------------------------------------------------------------------------------------
 30 | ERROR | Missing parameter comment
 31 | ERROR | Missing parameter comment
 32 | ERROR | Missing parameter comment
 33 | ERROR | Missing parameter comment
-------------------------------------------------------------------------------------------


FILE: /app/drupal8/web/modules/custom/nidirect_contacts/src/Controller/ContactListingController.php
---------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
---------------------------------------------------------------------------------------------------
 105 | ERROR | [x] Functions must not contain multiple empty lines in a row; found 2 empty lines
---------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
---------------------------------------------------------------------------------------------------
```

**Comment by johangant** [comment]
This PR provides a bunch of things:

- Breadcrumb builder service adjustment so it can strip the breadcrumb on views pages and show them on contact node pages.
- Replaces the present panels + views + custom breadcrumb solution with a simpler, route + controller and two views to drive the data extraction/presentation:
  1. Search API connected main search view: allows fuzzy searching and sorting.
  2. Node driven view: connected to the database. A bit annoying to need this, but it was far easier than trying to wrangle with Lucene syntax + switching off search value stemming to allow for single letter searches directly from Solr. In time, this can be replaced with a fully Solr-centric solution if a workaround can be found.
- A custom module that alters the exposed filter query behaviour in view 2 above, and preprocess hooks to show/control convenience links on the page depending on what you've searched for/enabled.
- Nightwatch.js UI tests to capture some of the fiddlier display behaviour that comes from certain states of the search.

NB: for testing, you'll need to copy the two .twig files in the templates folder into your active theme's corresponding templates folder.

File: nidirect_contacts/nidirect_contacts.info.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name: 'NIDirect Contacts'
+type: module
+description: 'Custom code to provide NIDirect contact search enhancements and forms'
+core: 8.x
+package: 'Custom'
```
Comment (omahm) [inline nidirect_contacts/nidirect_contacts.info.yml]: Package should be 'NIDirect'

File: nidirect_contacts/nidirect_contacts.module [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+function nidirect_contacts_preprocess_views_view(&$variables) {
+  if (!preg_match('/^contacts/', $variables['id'])) {
+    return;
+  }
+
+  if (!in_array($variables['display_id'], ['contact_search', 'contacts_by_letter'])) {
+    return;
+  }
+
+  $q = \Drupal::request()->query->all();
```
Comment (omahm) [inline nidirect_contacts/nidirect_contacts.module]: Not an issue, but there's the shorter Request class : https://api.drupal.org/api/drupal/vendor%21symfony%21http-foundation%21Request.php/class/Request/8.2.x

File: nidirect_contacts/src/Controller/ContactListingController.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+   *    <Big A-Z, 0-9 grid of links>
+   *    Link is: /contacts/letter/a
+   *
+   * @return array
+   *   Return render array for Drupal to convert to HTML.
+   */
+  public function default() {
+    $content = [];
+
+    // Views static service class used in absence of container service to inject via constructor.
```
Comment (omahm) [inline nidirect_contacts/src/Controller/ContactListingController.php]: Not sure how this comment is relevant to the render array?

File: nidirect_contacts/src/Plugin/Block/ContactAzBlock.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+   */
+  public function build() {
+    $build = [];
+    $links = [];
+
+    $title = $this->t('Find contacts beginning with') . '...';
+
+    if ($this->routeMatch->getRouteName() == 'nidirect_contacts.letter') {
+      $letter = $this->routeMatch->getParameter('letter');
+      $title = $this->t('Showing entries for') . ' ' . strtoupper($letter);
```
Comment (omahm) [inline nidirect_contacts/src/Plugin/Block/ContactAzBlock.php]: $letter should be passed into t() as a placeholder.

File: nidirect_contacts/src/Plugin/Block/ContactAzBlock.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    $build['title'] = [
+      '#markup' => '<p>' . $title . '</p>',
+    ];
+
+    foreach (array_merge(range('a', 'z'), range('0', '9')) as $item) {
+      $links[] = Link::createFromRoute(strtoupper($item), 'nidirect_contacts.letter', ['letter' => $item], [
+        'attributes' => [
+          'title' => $this->t('View entries under') . ' ' . strtoupper($item)
```
Comment (omahm) [inline nidirect_contacts/src/Plugin/Block/ContactAzBlock.php]: Pass as placeholder

File: nidirect_contacts/nidirect_contacts.info.yml (hunk 1 above)
Comment (johangant) [reply nidirect_contacts/nidirect_contacts.info.yml]: 👍

File: nidirect_contacts/nidirect_contacts.module (hunk 2 above)
Comment (johangant) [reply nidirect_contacts/nidirect_contacts.module]: Cool, thanks. The query parameters are likely accessible either way, I've followed what's typically done in core modules when in a .inc or .module context where they reference this static service.

File: nidirect_contacts/src/Controller/ContactListingController.php (hunk 3 above)
Comment (johangant) [reply nidirect_contacts/src/Controller/ContactListingController.php]: Me neither, and I wrote it! Putting it down to a brain-fart and slimming down the comment block with the next commit :)

File: nidirect_contacts/src/Plugin/Block/ContactAzBlock.php (hunk 4 above)
Comment (johangant) [reply nidirect_contacts/src/Plugin/Block/ContactAzBlock.php]: 👍

File: nidirect_contacts/src/Plugin/Block/ContactAzBlock.php (hunk 5 above)
Comment (johangant) [reply nidirect_contacts/src/Plugin/Block/ContactAzBlock.php]: 👍

**Pull request merged**